Write Reflection Perspective (#249)
ITU-DevOps2026/ITU-minitwit: report/sections/03-0-reflection_perspective.md

of your ITU-MiniTwit systems. Link back to respective commit messages, issues, tickets, etc. to illustrate these.

Also reflect and describe what was the "DevOps" style of your work. For example, what did you do differently to previous development projects and how did it work?
-->

## Multi-instance environment (Docker Swarm):

To implement Docker Swarm, the application had to be stateless. However, the "latest" value was originally stored in memory, so we refactored the system to persist the "latest" value in the database. 

> Another issue was the users login state. The loadbalancing introduced by both docker swarm and traefik caused our users to be logged out ocassinally as there requests were send to different application-instances each time. To fix this we added sticky session cookies, which routes request from a user to the same instance each time.

Reviewer: ```suggestion
Another issue was the users login state. The loadbalancing introduced by both docker swarm and traefik caused our users to be logged out occasionally as their requests were send to different application-instances each time. To fix this we added sticky session cookies, which routes request from a user to the same instance each time.
```

> Additionally, moving to Swarm broke our initial monitoring setup because Prometheus could not easily track dynamic service instances. Regarding monitoring, we reduced the fluctuations (cretaed by the inconsistency in the different instances) by regularly updating the monitoring metrics based on the database. This still only solved some of the problem. If we had more time, we belive the better solution would be to move the prometheus service to be a global service in our swarm and then show both monitoring of each instance and an average of all instances in grafana. 

Reviewer: ```suggestion
Additionally, moving to Swarm broke our initial monitoring setup because Prometheus could not easily track dynamic service instances. Regarding monitoring, we reduced the fluctuations (created by the inconsistency in the different instances) by regularly updating the monitoring metrics based on the database. This still only solved some of the problem. If we had more time, we believe the better solution would be to move the Prometheus service to be a global service in our swarm and then show both monitoring of each instance and an average of all instances in Grafana. 
```

We now know statelessness is a prerequisite for scalability and that the monitoring strategy must evolve concurrently to handle dynamic, multi-instance environments.

## The need for orchestration logging:

> Our current logs in grafana only shows application-logs (retrived using serilog and elasticsearch), but as our system matured with docker swarm, we found ourselves checking the droplet's logs via ssh very often. 

Reviewer: ```suggestion
Our current logs in Grafana only shows application-logs (retrieved using Serilog and Elasticsearch), but as our system matured with Docker Swarm, we found ourselves checking the droplet's logs via SSH very often. 
```

> The logs of the droplets would provide great value if they were visible in our grafana dashboards to keep all in one location. We imagine this could be done by adding a logshipper (we were considering filebeat) as a seperate service to send logs through elasticsearch. 

Reviewer: ```suggestion
The logs of the droplets would provide great value if they were visible in our Grafana dashboards to keep all in one location. If we had more time, we would have implemented this by adding a logshipper (we were considering Filebeat) as a separate service to send logs through Elasticsearch. 
```

We learned that application logs represent only half of the picture. Infrastructure and orchestration logging are equally important, and standard application-level logging libraries cannot substitute for proper node-level log shipping.

## Availability and Database Migration

When migrating our live production data from SQLite to MySQL we wanted minimal downtime and persistence of data integrity. We introduced an abstraction layer using EF Core. This decoupled our application logic from the database implementation, allowing us to swap providers with minimal code changes. But still, when testing moving the data in a single batch, it took roughly 10 minutes which would cause an unacceptable window of application downtime and data loss. Furthermore, we faced text-encoding anomalies (corrupted characters) and format incompatibilities during the dump transfer.

> We added utf8mb4 encoding to our schemas and cleaned SQLite-specific elements using a tailored pipeline (sed, stripping Byte Order Marks). To minimize downtime, we developed a multi-batch migration strategy: we made a large initial seed, and right before the switch, we used the Linux comm tool to calculate a dump consisting only of new records written during the sync.

Reviewer: ```suggestion
To handle this, we added utf8mb4 encoding to our schemas and cleaned SQLite-specific elements using a tailored pipeline (sed, stripping Byte Order Marks). To minimize downtime, we developed a multi-batch migration strategy: we made a large initial seed, and right before the switch, we used the Linux comm tool to calculate a dump consisting only of new records written during the sync.
```

During the final deployment, an omission in updating the production environment connection string caused a few minutes of unintended downtime, where we missed 19 register requests according to our dashboard tracking. This showed how even minor manual steps in configuration updates introduce human error, highlighting the necessity of automation.

## Attacks on the Database

> A few hours after migrating to MySQL, our database droplet was breached by automated ransomware bots because it was exposed to the public internet without a firewall, resulting in total data erasure and a Bitcoin ransom demand. To prevent this in the future, we used Vagrant to provision firewalls and configured MySQL to only accept connections from specific IP addresses (our application droplets).

Reviewer: ```suggestion
A few hours after migrating to MySQL, our database droplet was breached by automated ransomware bots because it was exposed to the public internet without a firewall, resulting in total data erasure and a Bitcoin ransom demand (see ransom note in the appendix [Ransom note from hacker](#ransom_note_from_hacker)). To prevent this in the future, we used Vagrant to provision firewalls and configured MySQL to only accept connections from specific IP addresses (our application droplets). Additionally we added a Digital ocean firewall, described in the section called [Security](#security).
```

We also experienced a large amount of automated SSH attacks on our database droplet, which would result in MiniTwit application being down for long periods while the database spent all its' CPU rejecting these login attempts. Therefore, we set up fail2ban on the database droplet, which puts IP addresses in "jail" for a day if they fail two login attempts in a row. This helped a lot, and currently we have around 200 IP addresses in jail, which shows us that these attacks are still ongoing but we now have a way to defer them. 

We now know that infrastructure must be hardened before it goes live, not as an afterthought and even if an attacker doesn't manage to get access to a droplet, they can still do a lot of damage by overwhelming the server, which is important to take into account when setting up security. 

## Limitations of Hardened Images 

> We experienced failing tests when running docker compose. We expected it to be caused by race conditions where the tests were running before the MiniTwit application was completely initialized. To prevent this, we wanted to add a simple health check that would curl the endpoint, but the curl tool can not be installed on Docker hardened images. Instead, we created a sidecar container that uses a curl image to curl the application and made the tests depend on this container. This was a relevant lesson on the limitations of hardened images and how to handle it.

Reviewer: ```suggestion
We experienced failing tests when running docker compose. We expected it to be caused by race conditions where the tests were running before the MiniTwit application was completely initialized. To prevent this, we wanted to add a simple health check that would curl the endpoint, but the curl tool can't be installed on Docker hardened images. Instead, we created a sidecar container that uses a curl image to curl the application and made the tests depend on this container. This was a relevant lesson on the limitations of hardened images and how to handle it.
```

## Resource Management and OOM Errors:

> As the simulator had been running for a while, the amount of data pilled up, which led to our application becoming very slow. We therefore indexed our data according to what we wanted to show and optimized our code avoiding subdependent queries. 

Reviewer: ```suggestion
As the simulator had been running for a while, the amount of data piled up, which led to our application becoming very slow. We therefore indexed our data according to what we wanted to show and optimized our code to avoid sub-dependent queries. 
```

> Our droplet kept crashing (exit code 137) because Elasticsearch consumed all available RAM. First of all, we moved all logging and monitoring to a seperat droplet than the application. We learned to tune the JVM heap size limiting it to 50% of the available memory on the droplet.

Reviewer: ```suggestion
Our droplet kept crashing (exit code 137) because Elasticsearch consumed all available RAM. To mitigate the problem, we moved all logging and monitoring to a seperate droplet than the application. We learned to tune the JVM heap size limiting it to 50% of the available memory on the droplet.
```

> Our droplet kept crashing (exit code 137) because Elasticsearch consumed all available RAM. First of all, we moved all logging and monitoring to a seperat droplet than the application. We learned to tune the JVM heap size limiting it to 50% of the available memory on the droplet.

Reviewer 2: ```suggestion
Our droplet kept crashing (exit code 137) because Elasticsearch consumed all available RAM. First of all, we moved all logging and monitoring to a separate droplet than the application. We learned to tune the JVM heap size limiting it to 50% of the available memory on the droplet.
```

For emergencies, we also implemented a swap file to take over processes if we run out of RAM.

> We realized that our 25GB disk space on the logging droplet would quickly fill up with logs so we configured Index Lifecycle Management (ILM) rules within Elasticsearch to enforce an explicit rollover policy, ensuring older indices are automatically pruned and deleted after reaching age thresholds.

Reviewer: ```suggestion
We realized that our 25GB disk space on the logging droplet would quickly fill up with logs so we configured Index Lifecycle Management (ILM) rules within Elasticsearch to enforce an explicit rollover policy, ensuring older indices are automatically pruned and deleted after reaching age thresholds.(which is 14 days in our case)
```

Author: Don't think we need to say this

## DevOps Style of Work
In this project, our DevOps style was defined by the "Three Ways" of DevOps:

- __Flow__: Unlike previous projects where we might have worked in long-lived branches, we focused on small batch sizes and a deployment pipeline ensuring value gets to prod quickly. We used GitHub Actions to ensure that code was tested and deployed to "Test" and "Prod" environments automatically. The use of vagrant and docker ensures environment consistency. 

- __Feedback__: We implemented monitoring and logging and added downtime/latency email alerts. This gave us immediate feedback on whether a deployment was successful or if the system was under stress. 
We also included tests and static code analysis tools in our pipeline telling us exactly which part of our code had issues so they could be resolved quickly. 

> - __Continuous Learning and Experimentation__: We followed a blameless culture with room for mistakes and focus on understanding the issues. We maintained a weekly logbook to document considerations and decisions and regularly updated readmes with guides on running, deploying and etc. This ensured that knowledge was shared across the team, and everyone had access to the information they needed to contribute effectively.

Reviewer 2: ```suggestion
- __Continuous Learning and Experimentation__: We followed a blameless culture with room for mistakes and focus on understanding the issues. We maintained a weekly logbook to document considerations and decisions and regularly updated ReadMes with guides on running, deploying and etc. This ensured that knowledge was shared across the team, and everyone had access to the information they needed to contribute effectively.
```
